src/SourcemapToolkit.SourcemapParser/SourceMap.cs

```diff
@@ -1,51 +1,133 @@
+        /// </summary>
+        public int Version;
+
+        /// <summary>
+        /// The name of the generated file to which this source map corresponds
+        /// </summary>
+        public string File;
+
+        /// <summary>
+        /// The raw, unparsed mappings entry of the soure map
+        /// </summary>
+        public string Mappings;
+
+        /// <summary>
+        /// The list of source files that were the inputs used to generate this output file
+        /// </summary>
+        public List<string> Sources;
+
+        /// <summary>
+        /// A list of known original names for entries in this file
+        /// </summary>
+        public List<string> Names;
+
+        /// <summary>
+        /// Parsed version of the mappings string that is used for getting original names and source positions
+        /// </summary>
+        public List<MappingEntry> ParsedMappings;
+
+        /// <summary>
+        /// Applies the mappings of a sub source map to the current source map
+        /// Each mapping to the supplied source file is rewritten using the supplied source map
+        /// <param name="submap">The submap to apply</param>
+        /// <param name="aSourceFile">The filename of the source file. If not specified, submap's File property will be used</param>
+        /// <returns>A new source map</returns>
+        /// </summary>
+        public SourceMap ApplySourceMap(SourceMap submap, string aSourceFile = null)
+        {
+            if (submap == null)
+            {
+                throw new ArgumentNullException(nameof(submap));
+            }
+
+            string sourceFile = aSourceFile;
+            if (sourceFile == null)
+            {
+                if (submap.File == null)
+                {
+                    throw new Exception("ApplySourceMap expects either the explicit source file to the map, or submap's 'file' property");
+                }
+
+                sourceFile = submap.File;
+            }
+
+            List<string> newSources = new List<string>();
+            List<string> newNames = new List<string>();
+            List<MappingEntry> newMappingEntries = new List<MappingEntry>();
+
+            SourceMap newSourceMap = new SourceMap
+            {
+                File = this.File,
+                Version = this.Version,
+                Sources = newSources,
+                Names = newNames,
+                ParsedMappings = newMappingEntries
+            };
+
+            // transform mappings in this source map
+            foreach (MappingEntry mappingEntry in this.ParsedMappings)
+            {
+                MappingEntry newMappingEntry = mappingEntry.Clone() as MappingEntry;
+                SourcePosition original = mappingEntry.OriginalSourcePosition;
+
+                if (mappingEntry.OriginalFileName == sourceFile && original != null)
+                {
+                    MappingEntry correspondingEntry = submap.GetMappingEntryForGeneratedSourcePosition(original);
```

Comment:
it goes backwards, so we would do `bc.sourcemap.js.apply(ab.sourcemap.js)` to get `ca.sourcemap.js`

example: a.js -> b.js -> c.js
ab.map contains: (line 3: col 5) to (line 2: col 4), which points generated (3:5) in b.js to original (2:4) in a.js
bc.map contains: (line 4: col 3) to (line 3: col 5), which points generated (4:3) in c.js to original (3:5) in b.js

so the apply function looks at (line 4: col 3), finds that it maps to **original source position** (line 3: col 5), and finds the entry for the **generated source position** in `ab.map` with (line 3: col 5). Which is (line 3: col 5) to (line 2: col 4). So we add a mapping entry from (line 4: col 3) to (line 2: col: 4)